### PR TITLE
fathead metadata

### PR DESCRIPTION
**Pull request by jdorweiler**
Adds `source_data` to access fathead specific metadata and a hash indexed by source id.
@zachthompson @russellholt

### CODE REVIEW

File: lib/DDG/Meta/Data.pm [hunk 1]
```diff
@@ -80,6 +80,8 @@ unless(%ia_metadata){
             $ia_metadata{id}{$id} = $module_data;
             #add new ia to ia_metadata{module}. Multiple ias per module possible
             push @{$ia_metadata{module}{$perl_module}}, $module_data;
+            # by source number for fatheads
+            $ia_metadata{source}{$module_data->{src_id}} = $module_data if $module_data->{repo} eq 'fathead';
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Set a variable once, e.g. $is_fathead, outside of this loop using $type.

File: lib/DDG/Meta/Data.pm [hunk 2]
```diff
@@ -167,4 +169,18 @@ sub _js_callback_name {
     return $fn;
 }
 
+# return fathead data
+sub source_data {
+    my ($self, $source) = @_;
+    my $data = $ia_metadata{source}{$source} if exists $ia_metadata{source}{$source};
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: If(my $data = $ia_metadata{source}{$source}){
    @return = ....
}

File: lib/DDG/Meta/Data.pm [hunk 3]
```diff
@@ -167,4 +169,18 @@ sub _js_callback_name {
...
+    my @return = (
+        $data->{src_name},
+        $data->{src_domain},
+        $data->{src_options}->{source_info},
+        $data->{src_options}->{is_mediawiki},
+        $data->{perl_module}
+    );
+    return @return;
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Return a reference or 0 if not found.

Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: I'm supporting the old interface.  The caller wants a list.

Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: It's still a list, just a reference.  You can do 
``` Perl
my @expecting_list = @{DDG::Meta::Data->source_data($src)};
```
if you need it to be a list immediately.

File: lib/DDG/Meta/Data.pm (hunk 2 above)
Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: need to return an array of empty strings if nothing is found.

File: lib/DDG/Meta/Data.pm [hunk 4]
```diff
@@ -162,9 +174,13 @@ sub get_ia {
...
 sub get_js {
-    my ($self, $id) = @_;
+    my ($self, $by, $lookup) = @_;
+    my $ia = $self->get_ia($by => $lookup);
+    return unless $ia;
 
-    my $metaj = eval { JSON::XS->new->ascii->encode($self->get_ia(id => $id)) } || qq|{"encode_json error":"$@"}|;
+    my $metaj;
+    my $id = $ia->{id};
+    eval { $metaj = JSON::XS->new->ascii->encode($ia) } || return;
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Move my $meta here and change || to or.

File: lib/DDG/Meta/Data.pm [hunk 5]
```diff
@@ -185,4 +201,36 @@ sub _js_callback_name {
...
+    }else{
+        return ('','','','','');
+    }
+}
+
+sub source_skip {
+    my ($self, $source) = @_;
+    if(my $data = $ia_metadata{source}{$source}){
+        return (
+            $data->{src_options}->{source_name} || '',
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: my $src_options = $data->{src_options}?

File: lib/DDG/Meta/Data.pm [hunk 6]
```diff
@@ -185,4 +201,36 @@ sub _js_callback_name {
     return $fn;
 }
 
+# return fathead data
+sub source_data {
+    my ($self, $source) = @_;
+    if(my $data = $ia_metadata{source}{$source}){
+        return (
```
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: Can the code using these functions not handle an empty array?  This really should be:
``` Perl
my @d;
if(my $ia = ...){
    @d = (
    );
}
return @d;
```
This empty string stuff shouldn't be necessary, unless you're getting undefined errors?

Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: There will be undefined errors so it actually needs an array of empty strings.

File: lib/DDG/Meta/Data.pm (hunk 4 above)
Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: It's used below so needs to be declared outside the eval?

File: lib/DDG/Meta/Data.pm (hunk 5 above)
Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: I don't get it

File: lib/DDG/Meta/Data.pm (hunk 4 above)
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: It was before
``` Perl
my $metaj = eval { ....
```
not sure why that changed.  Also, if we're going to allow $by to be 'module' now, you need to check for $ia being an array.   Do we really need to allow this?

File: lib/DDG/Meta/Data.pm (hunk 5 above)
Comment (zachthompson) [inline lib/DDG/Meta/Data.pm]: $src_options->{skip_end},
etc.

File: lib/DDG/Meta/Data.pm (hunk 4 above)
Comment (jdorweiler) [inline lib/DDG/Meta/Data.pm]: I really just needed to look up by source for fatheads.  Maybe we just support look up by id and source for now.

**Comment by russellholt** [comment]
:+1:

**Pull request merged**